src/scf/newton.jl

```diff
@@ -0,0 +1,360 @@
+    count = mpi_sum(count, basis.comm_kpts)
+    G_to_r(basis, δρ_fourier) ./ count
+end
+
+"""
+    apply_K(basis::PlaneWaveBasis, δψ, ψ, ρ, occupation)
+
+Compute the application of K to an element in the space tangent to ψ.
+"""
+function apply_K(basis::PlaneWaveBasis, δψ, ψ, ρ, occupation)
+    Nk = length(basis.kpoints)
+
+    δψ = proj_tangent(δψ, ψ)
+    dρ = compute_dρ(basis, ψ, δψ, ρ, occupation)
+    dV = apply_kernel(basis, dρ; ρ=ρ)
+
+    Kδψ = []
+    for ik = 1:Nk
+        kpt = basis.kpoints[ik]
+        ψk = ψ[ik]
+        dVψk = similar(ψk)
+
+        for i = 1:size(ψk, 2)
+            ψk_r = G_to_r(basis, kpt, ψk[:,i])
+            dVψk_r = dV[:, :, :, kpt.spin] .* ψk_r
+            dVψk[:, i] = r_to_G(basis, kpt, dVψk_r)
+        end
+        push!(Kδψ, dVψk)
+    end
+    # ensure proper projection onto the tangent space
+    proj_tangent!(Kδψ, ψ)
+end
+
+# Callback and convergence
+
+"""
+Flag convergence as soon as density change drops below tolerance
+"""
+function NewtonConvergenceDensity(tolerance)
+    info -> (norm(info.ρ_next - info.ρ) * sqrt(info.basis.dvol) < tolerance)
+end
+
+"""
+Default callback function for `newton`, which prints a convergence table
+"""
+function NewtonDefaultCallback()
+    prev_energies = nothing
+    function callback(info)
+        !mpi_master() && return info  # Printing only on master
+        if info.stage == :finalize
+            info.converged || @warn "Newton not converged."
+            return info
+        end
+        collinear = info.basis.model.spin_polarization == :collinear
+
+        if info.n_iter == 1
+            E_label = haskey(info.energies, "Entropy") ? "Free energy" : "Energy"
+            magn    = collinear ? ("   Magnet", "   ------") : ("", "")
+            @printf "n     %-12s      Eₙ-Eₙ₋₁     ρ_next-ρ%s\n" E_label magn[1]
+            @printf "---   ---------------   ---------   --------%s\n" magn[2]
+        end
+        E    = isnothing(info.energies) ? Inf : info.energies.total
+        Δρ   = norm(info.ρ_next - info.ρ) * sqrt(info.basis.dvol)
+        if size(info.ρ_next, 4) == 1
+            magn = NaN
+        else
+            magn = sum(spin_density(info.ρ_next)) * info.basis.dvol
+        end
+
+        Estr   = (@sprintf "%+15.12f" round(E, sigdigits=13))[1:15]
+        prev_E = prev_energies === nothing ? Inf : prev_energies.total
+        ΔE     = prev_E == Inf ? "      NaN" : @sprintf "% 3.2e" E - prev_E
+        Mstr = collinear ? "   $((@sprintf "%6.3f" round(magn, sigdigits=4))[1:6])" : ""
+        @printf "% 3d   %s   %s   %2.2e%s\n" info.n_iter Estr ΔE Δρ Mstr
+        prev_energies = info.energies
+
+        flush(stdout)
+        info
+    end
+end
+
+# Newton algorithm
+
+"""
+    newton_step(basis::PlaneWaveBasis, ψ, res, occupation;
+                tol_cg=1e-10, verbose=false)
+
+Perform a Newton step : we take as given a planewave set ψ and we return the
+Newton step δψ and the updated state ψ - δψ
```

Comment:
Hm, I don't like the convention you have of psi - deltapsi. Can we switch to psi + deltapsi, here and in the comment at the top? So Newton is xn+1 = x + dx, where J dx = - res

Reply:
Ok, but then I suggest to let `newton_step` as it is, otherwise it will be confusing that newton_step(..., res) actually solves (O+K)x= -res, so we just pass -res as argument to `newton_stepè.

Reply:
That's even more confusing because it introduces two different ocnventions for what "res" means. Just rename newton_step to solve_Omega_plus_K(psi, rhs)? It's better also for what we want to do for AD.

Reply:
Yup, sounds better !